aws/resource_aws_s3_bucket.go

```diff
@@ -2777,13 +2784,76 @@ func flattenS3ObjectLockConfiguration(conf *s3.ObjectLockConfiguration) []interf
+	for _, granteeObject := range ap.Grants {
+		if aws.StringValue(granteeObject.Grantee.ID) == aws.StringValue(ap.Owner.ID) &&
+			aws.StringValue(granteeObject.Permission) == s3.PermissionFullControl {
+			owner_full_control = true
+			continue
+		}
+		if aws.StringValue(granteeObject.Grantee.URI) == "http://acs.amazonaws.com/groups/global/AllUsers" &&
+			aws.StringValue(granteeObject.Permission) == s3.PermissionRead {
+			all_users_read = true
+			continue
+		}
+		if aws.StringValue(granteeObject.Grantee.URI) == "http://acs.amazonaws.com/groups/global/AllUsers" &&
+			aws.StringValue(granteeObject.Permission) == s3.PermissionWrite {
+			all_users_write = true
+			continue
+		}
+		if aws.StringValue(granteeObject.Grantee.URI) == "http://acs.amazonaws.com/groups/global/AuthenticatedUsers" &&
+			aws.StringValue(granteeObject.Permission) == s3.PermissionRead {
+			authenticated_users_read = true
+			continue
+		}
+		// This is the Amazon ID for the za-team group used for EC2 operations in S3 buckets
+		if aws.StringValue(granteeObject.Grantee.ID) == "6aa5a366c34c1cbe25dc49211496e913e0351eb0e8c37aa3477e40942ec6b97c" &&
+			aws.StringValue(granteeObject.Permission) == s3.PermissionRead {
+			ec2_read = true
+			continue
+		}
+		if aws.StringValue(granteeObject.Grantee.URI) == "http://acs.amazonaws.com/groups/s3/LogDelivery" &&
+			aws.StringValue(granteeObject.Permission) == s3.PermissionReadAcp {
+			logdelivery_read_acp = true
+			continue
+		}
+		if aws.StringValue(granteeObject.Grantee.URI) == "http://acs.amazonaws.com/groups/s3/LogDelivery" &&
+			aws.StringValue(granteeObject.Permission) == s3.PermissionWrite {
+			logdelivery_write = true
+			continue
+		}
+
+		// This is a non standard grant combination, so it's not a canned ACL
+		return false
 	}
 
+	switch {
+	case owner_full_control && !all_users_read && !all_users_write && !authenticated_users_read && !ec2_read && !logdelivery_read_acp && !logdelivery_write:
+		d.Set("acl", "private")
+	case owner_full_control && all_users_read && !all_users_write && !authenticated_users_read && !ec2_read && !logdelivery_read_acp && !logdelivery_write:
+		d.Set("acl", "public-read")
+	case owner_full_control && all_users_read && all_users_write && !authenticated_users_read && !ec2_read && !logdelivery_read_acp && !logdelivery_write:
+		d.Set("acl", "public-read-write")
+	case owner_full_control && !all_users_read && !all_users_write && authenticated_users_read && !ec2_read && !logdelivery_read_acp && !logdelivery_write:
+		d.Set("acl", "authenticated-read")
+	case owner_full_control && !all_users_read && !all_users_write && !authenticated_users_read && ec2_read && !logdelivery_read_acp && !logdelivery_write:
+		d.Set("acl", "aws-exec-read")
+	case owner_full_control && !all_users_read && !all_users_write && !authenticated_users_read && !ec2_read && logdelivery_read_acp && logdelivery_write:
+		d.Set("acl", "log-delivery-write")
+	}
```

Comment:
We probably want to add a `default` branch that returns `false` to indicate that no canned ACL matched and that we need to record all grants.

Reply:
Added a [test case](https://github.com/hashicorp/terraform-provider-aws/blob/d0d121f12f2c593d543d443915c4c187707f080f/aws/resource_aws_s3_bucket_test.go#L762-L784) for this:
```
make testacc TESTARGS='-run=TestAccAWSS3Bucket_Security_updateGrant'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSS3Bucket_Security_updateGrant -timeout 180m
=== RUN   TestAccAWSS3Bucket_Security_updateGrant
=== PAUSE TestAccAWSS3Bucket_Security_updateGrant
=== CONT  TestAccAWSS3Bucket_Security_updateGrant
    resource_aws_s3_bucket_test.go:712: Step 5/8 error: After applying this test step and performing a `terraform refresh`, the plan was not empty.
        stdout


        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place

        Terraform will perform the following actions:

          # aws_s3_bucket.bucket will be updated in-place
          ~ resource "aws_s3_bucket" "bucket" {
                id                          = "tf-test-bucket-42309041846304674"
                tags                        = {}
                # (10 unchanged attributes hidden)

              + grant {
                  + permissions = [
                      + "READ",
                    ]
                  + type        = "Group"
                  + uri         = "http://acs.amazonaws.com/groups/global/AllUsers"
                }
              + grant {
                  + permissions = [
                      + "READ_ACP",
                    ]
                  + type        = "Group"
                  + uri         = "http://acs.amazonaws.com/groups/s3/LogDelivery"
                }
              + grant {
                  + id          = "d6d601aee664a28898ebdc60021fbb05370bb7147e4547a6b53992e82e04d96d"
                  + permissions = [
                      + "FULL_CONTROL",
                    ]
                  + type        = "CanonicalUser"
                }

                # (1 unchanged block hidden)
            }

        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccAWSS3Bucket_Security_updateGrant (170.02s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-aws/aws       170.104s
FAIL
make: *** [GNUmakefile:30: testacc] Error 1
```

And implemented the above mentioned fix to make the test pass:

```
# make testacc TESTARGS='-run=TestAccAWSS3Bucket_Security_updateGrant'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSS3Bucket_Security_updateGrant -timeout 180m
=== RUN   TestAccAWSS3Bucket_Security_updateGrant
=== PAUSE TestAccAWSS3Bucket_Security_updateGrant
=== CONT  TestAccAWSS3Bucket_Security_updateGrant
--- PASS: TestAccAWSS3Bucket_Security_updateGrant (242.17s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       242.255s
```